XRPLF/rippled
# Add the ability to mark amendments as obsolete

## High Level Overview of Change

* Prevents the server from ever voting on the given amendment,
  regardless of operator configuration.
* Also prevents the "feature" command from changing the amendment's
  vote.

### Context of Change

This comment: https://github.com/XRPLF/rippled/pull/4282#issuecomment-1223776779 pointed out an ongoing issue - there are amendments that simply shouldn't be voted on because they're broken, there's no actual code implementing them, or other reasons. Unfortunately, they need to remain supported on the main net servers on the off chance that they do get enabled. Remember: If they get enabled, any versions that don't support them will be amendment blocked.

Up to now, the communication to let operators know not to vote for these amendments has been done via back channels, blog posts, etc. Why not use code to both communicate and enforce this message?

A determined operator is still not absolutely prevent from voting against an amendment. They are free to change the source and compile locally, and still stick to the rules of the protocol.

### Type of Change


## Before / After

Before: there are two options for an amendment's `DefaultVote` behaviour: yes and no.

After: there are three options for an amendments `VoteBehavior`: DefaultYes, DefaultNo, and Obsolete.

## Test Plan

There are currently four amendments marked `Obsolete` that haven't passed yet:
1. `CryptoConditionsSuite`
2. `NonFungibleTokensV1`
3. `fixNFTokenDirV1`
4. `fixNFTokenNegOffer`

Test cases:
1. Try voting for them. Betcha can't.
2. Run this code against a test network running majority older versions. Get one of these amendments enabled. Verify that this code doesn't get amendment blocked.

## Future Tasks

Amendment governance is an ongoing discussion. I am not going to rehash all the details here.
